### PR TITLE
Add fully configurable activation checkpointing

### PR DESCRIPTION
**Pull request by cli99**
This PR allows (when `activation_checkpointing: true`) specifying activation checkpointing target `activation_checkpointing_target` as :
            1. null (or no such field), then ack ckpt the whole MPTBlock on all layers
            2. a list of modules to act ckpt, e.g.,
                activation_checkpointing_target:
                    - grouped_query_attention
                    - mptmlp
                , then ack ckpt all the modules in the list on all layers
            3. a dictionary of module name with target_blocks, e.g.,
                activation_checkpointing_target:
                    {
                            "mptblock": target_blocks_1,
                            "grouped_query_attention": target_blocks_2
                    }
                target_blocks (target_blocks_1, target_blocks_2 above) can be:
                - a single integer n means the first n transformer block will be candidates for act ckpt
                - a string of first-n, middle-m, last-k, range-i-j means the first n, the middle m,  the last k, or the range [i, j) layers are candidates for act ckpt. E.g, 'first-2, last-2' means the first 2 and last 2 transformer blocks are candidates for act ckpt.
                    middle-m is range [start, end) where ``start = max(max_block_idx // 2 - m // 2, 0), end = min(start + m, max_block_idx + 1)``
                - a list of integers corresponds to the list of transformer block ids, e.g., [2] means the second transformer block is the candidate for act ckpt. [2, 3] means the second and third transformer block are candidates.
                - a list of mixed integers and strings of first-n, middle-m, last-k, range-i-j

            An example in yaml config file:,
                fsdp_config:
                    activation_checkpointing: true
                model:
                    activation_checkpointing_target:
                        {
                            "mptblock": 'first-5',
                            "grouped_query_attention": 'last-35'
                        }
                does full act ckpt on the first 5 layers and then ack ckpt the grouped_query_attention on the last 35 layers

### CODE REVIEW

File: llmfoundry/models/mpt/modeling_mpt.py [hunk 1]
```diff
@@ -908,41 +923,80 @@ def fsdp_wrap_fn(self, module: nn.Module) -> bool:
...
+                mod = get_act_ckpt_module(target)
+                act_ckpt_mod_to_blocks[mod] = -1
+        elif isinstance(act_ckpt_target, dict):
+            for k, v in act_ckpt_target.items():
+                mod = get_act_ckpt_module(k)
+                block_ids = get_target_block_list(v, module.max_block_idx)
+                act_ckpt_mod_to_blocks[mod] = block_ids
                 log.info(
-                    'Activation checkpointing MPTBlock only (ignoring other sub-block modules specified in activation_checkpointing_target).'
+                    f'for module {mod.__name__}, target_blocks is set as {v}, activation checkpointing is applied to {block_ids} blocks.'
```
Comment (mvpatel2000) [inline llmfoundry/models/mpt/modeling_mpt.py:984]:
```suggestion
                    f'For module {mod.__name__}, target_blocks is set as {v} so activation checkpointing is applied to {block_ids} blocks.'
```

File: llmfoundry/models/mpt/modeling_mpt.py [hunk 2]
```diff
@@ -352,6 +351,22 @@ def __init__(self, config: MPTConfig):
                 **config.to_dict(),
             ) for _ in range(config.n_layers)
         ])
+
+        def pass_on_block_idx(parent: nn.Module):
+            if not hasattr(parent, 'block_idx') or not hasattr(
+                    parent, 'max_block_idx'):
+                return
```
[the remark this answers is not in the document]
Comment (cli99) [reply llmfoundry/models/mpt/modeling_mpt.py:358]: yes, only mptblock and everything inside need to be tagged

File: llmfoundry/models/mpt/modeling_mpt.py [hunk 3]
```diff
@@ -908,41 +923,80 @@ def fsdp_wrap_fn(self, module: nn.Module) -> bool:
...
+                f'No activating checkpointing for {module.__class__.__name__}, only transformer block or its submodules are eligible for activation checkpointing.'
             )
-
-        if 'MPTBlock' in act_ckpt_list or 'mptblock' in act_ckpt_list:
-            if len(act_ckpt_list) > 1:
+            return False
+
+        act_ckpt_target = getattr(self.config,
+                                  'activation_checkpointing_target', None)
+        act_ckpt_mod_to_blocks = {}
```
[the remark this answers is not in the document]
Comment (cli99) [reply llmfoundry/models/mpt/modeling_mpt.py:965]: activation_checkpointing_fn is invoked on all the modules. If you want to do it once in the init (suppose in the MPTModel init), then we need to pass it along to all the submodules init. The downside is having messier init functions.

which part do you suggest to be a helper fn?

[the remark this answers is not in the document]
Comment (mvpatel2000) [reply llmfoundry/models/mpt/modeling_mpt.py:957]: Building `act_ckpt_mod_to_blocks`?

**Review by mvpatel2000 — APPROVED** [review]

**Pull request merged**